sdk-core/src/main/java/care/data4life/sdk/network/model/RecordEncryptionService.kt

```diff
@@ -0,0 +1,331 @@
+    }
+
+    override fun <T : Any> fromResource(
+        resource: T,
+        annotations: Annotations
+    ): NetworkModelContract.DecryptedBaseRecord<T> {
+        val tags = taggingService.appendDefaultTags(resource, null)
+
+        guard.checkTagsAndAnnotationsLimits(tags, annotations)
+
+        return buildRecord(
+            resource = resource,
+            tags = tags,
+            annotations = annotations,
+            creationDate = dateTimeFormatter.now(),
+            dataKey = cryptoService.generateGCKey().blockingGet()
+        )
+    }
+
+    private fun fetchCommonKey(): Pair<GCKey, String> {
+        return Pair(
+            cryptoService.fetchCurrentCommonKey(),
+            cryptoService.currentCommonKeyId
+        )
+    }
+
+    private fun encryptKey(
+        commonKey: GCKey,
+        mode: KeyType,
+        key: GCKey
+    ): EncryptedKey {
+        return cryptoService.encryptSymmetricKey(
+            commonKey,
+            mode,
+            key
+        ).blockingGet() as EncryptedKey
+    }
+
+    private fun encryptKeys(
+        commonKey: GCKey,
+        dataKey: GCKey,
+        attachmentKey: GCKey?
+    ): Pair<EncryptedKey, EncryptedKey?> {
+        val encryptedDataKey = encryptKey(commonKey, KeyType.DATA_KEY, dataKey)
+
+        return if (attachmentKey is GCKey) {
+            Pair(
+                encryptedDataKey,
+                encryptKey(commonKey, KeyType.ATTACHMENT_KEY, attachmentKey)
+            )
+        } else {
+            Pair(encryptedDataKey, null)
+        }
+    }
+
+    override fun <T : Any> encrypt(
+        decryptedRecord: NetworkModelContract.DecryptedBaseRecord<T>
+    ): NetworkModelContract.EncryptedRecord {
+        val dataKey = decryptedRecord.dataKey
+        val (commonKey, commonKeyId) = fetchCommonKey()
+        val (encryptedDataKey, encryptedAttachmentKey) = encryptKeys(
+            commonKey,
+            dataKey,
+            decryptedRecord.attachmentsKey
+        )
+
+        return EncryptedRecord(
+            commonKeyId,
+            decryptedRecord.identifier,
+            tagEncryptionService.encryptTagsAndAnnotations(
+                decryptedRecord.tags,
+                decryptedRecord.annotations
+            ),
+            fhirService._encryptResource(dataKey, decryptedRecord.resource),
+            decryptedRecord.customCreationDate,
+            encryptedDataKey,
+            encryptedAttachmentKey,
+            decryptedRecord.modelVersion
+        )
+    }
+
+    private fun validateRecord(encryptedResource: String, version: Int) {
+        if (encryptedResource.isEmpty()) {
+            throw CoreRuntimeException.InternalFailure()
+        }
+
+        if (!modelVersion.isModelVersionSupported(version)) {
+            throw DataValidationException.ModelVersionNotSupported("Please update SDK to latest version!")
+        }
+    }
+
+    private fun getCommonKey(commonKeyId: String, userId: String): GCKey {
```

Comment:
Naming is a bit confusing to me, we have 
`fetchCommonKey()` which returns a pair of GCKey and id
`getCommonKey()` which returns a GCKey 

and i wouldnt know which one refers to which by reading method names

Reply:
Have suggestions for better names?

Reply:
On a second look...and since its technically a private method...I guess it maps to the `getCommonKey` method you mentioned. Sure, the signature is different in terms of the parameter but the name and return value matches.